plutus-tx/src/Language/PlutusTx/Lift.hs

```diff
@@ -16,90 +16,111 @@ module Language.PlutusTx.Lift (
-    -- note: we typecheck&compile the plutus-tx term inside an empty builtin context (PLC.defConfig)
-    compiled <- flip runReaderT defaultCompilationCtx $ compileTerm True lifted
+    tcConfig <- PLC.getDefTypeCheckConfig $ Original ()
+    compiled <- flip runReaderT (toDefaultCompilationCtx tcConfig) $ compileTerm True lifted
     pure $ void $ UPLC.erase compiled
 
 -- | Get a Plutus Core program corresponding to the given value.
 safeLiftProgram
-    :: (Lift.Lift uni a
-       , PIR.AsTypeError e (PIR.Term TyName Name uni ()) uni (Provenance ()), PLC.GShow uni, PLC.GEq uni
+    :: (Lift.Lift uni fun a
+       , PIR.AsTypeError e (PIR.Term TyName Name uni fun ()) uni fun (Provenance ()), PLC.GShow uni, PLC.GEq uni
        , PIR.AsTypeErrorExt e uni (Provenance ())
-       , PLC.DefaultUni PLC.<: uni
-       , AsError e uni (Provenance ()), MonadError e m, MonadQuote m)
-    => a -> m (UPLC.Program Name uni ())
+
+       , AsError e uni fun (Provenance ()), MonadError e m, MonadQuote m
+       , PLC.ToBuiltinMeaning uni fun
+       )
+    => a -> m (UPLC.Program Name uni fun ())
 safeLiftProgram x = UPLC.Program () (PLC.defaultVersion ()) <$> safeLift x
 
 safeLiftCode
-    :: (Lift.Lift uni a
-       , PIR.AsTypeError e (PIR.Term TyName Name uni ()) uni (Provenance ()), PLC.GShow uni, PLC.GEq uni       , PLC.DefaultUni PLC.<: uni
+    :: (Lift.Lift uni fun a
+       , PIR.AsTypeError e (PIR.Term TyName Name uni fun ()) uni fun (Provenance ()), PLC.GShow uni, PLC.GEq uni
        , PIR.AsTypeErrorExt e uni (Provenance ())
-       , AsError e uni (Provenance ()), MonadError e m, MonadQuote m)
-    => a -> m (CompiledCode uni a)
+       , AsError e uni fun (Provenance ()), MonadError e m, MonadQuote m
+       , PLC.ToBuiltinMeaning uni fun
+       )
+    => a -> m (CompiledCode uni fun a)
 safeLiftCode x = DeserializedCode <$> safeLiftProgram x <*> pure Nothing
 
-unsafely :: Throwable uni => ExceptT (Error uni (Provenance ())) Quote a -> a
+unsafely
+    :: (Throwable uni, GHC.Typeable fun, Pretty fun)
+    => ExceptT (Error uni fun (Provenance ())) Quote a -> a
 unsafely ma = runQuote $ do
     run <- runExceptT ma
     case run of
         Left e  -> throw e
         Right t -> pure t
 
 -- | Get a Plutus Core term corresponding to the given value, throwing any errors that occur as exceptions and ignoring fresh names.
-lift :: (Lift.Lift uni a, Throwable uni) => a -> UPLC.Term Name uni ()
+lift
+    :: ( Lift.Lift uni fun a, Throwable uni
+       , GHC.Typeable fun, Pretty fun, PLC.ToBuiltinMeaning uni fun
+       )
+    => a -> UPLC.Term Name uni fun ()
 lift a = unsafely $ safeLift a
 
 -- | Get a Plutus Core program corresponding to the given value, throwing any errors that occur as exceptions and ignoring fresh names.
-liftProgram :: (Lift.Lift uni a, Throwable uni) => a -> UPLC.Program Name uni ()
+liftProgram
+    :: ( Lift.Lift uni fun a, Throwable uni
+       , GHC.Typeable fun, Pretty fun, PLC.ToBuiltinMeaning uni fun
```

Comment:
@michaelpj  And perhaps `Lift.Lift uni fun a` should imply `PLC.ToBuiltinMeaning uni fun`, so that we don't need to write that additional constraint everywhere compulsively?

Reply:
Since I removed `fun` from `Lift` and now it's only `Lift uni a`, it no longer makes any sense to make `PLC.ToBuiltinMeaning uni fun` a superclass.